### PR TITLE
gh-110652: Fix pre-commit hooks on macOS

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-110652

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ repos:
     hooks:
       - id: python-file-whitespace
         name: "Check Python file whitespace"
-        entry: 'python Tools/patchcheck/reindent.py --nobackup --newline LF'
+        entry: './Tools/patchcheck/reindent.py --nobackup --newline LF'
```
Comment (hugovk) [inline .pre-commit-config.yaml]: Also works without the `./`, any preference?

```suggestion
        entry: 'Tools/patchcheck/reindent.py --nobackup --newline LF'
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -36,7 +36,7 @@ repos:
     hooks:
       - id: c-file-whitespace
         name: "Check C file whitespace"
-        entry: "python Tools/patchcheck/untabify.py"
+        entry: "./Tools/patchcheck/untabify.py"
```
Comment (hugovk) [inline .pre-commit-config.yaml]:
```suggestion
        entry: "Tools/patchcheck/untabify.py"
```

**Comment by AlexWaygood** [comment]
This approach doesn't work on Windows, unfortunately. I think I'd prefer using a `python`-language hook, as @hugovk suggested in https://github.com/python/cpython/pull/109891#issuecomment-1755709823 (but see @AA-Turner's comments in https://github.com/python/cpython/pull/109891#issuecomment-1755755952)

**Review by hugovk — COMMENTED** [review]
Does this work on Windows?

cc @AA-Turner @AlexWaygood

File: .pre-commit-config.yaml (hunk 1 above)
Comment (erlend-aasland) [reply .pre-commit-config.yaml]: It's cleaner without the `./`

**Comment by AlexWaygood** [comment]
> Does this work on Windows?

no:

```
>pre-commit run --all-files
Run Ruff on Lib/test/....................................................Passed
Run Ruff on Argument Clinic..............................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
Check Python file whitespace.............................................Failed
- hook id: python-file-whitespace
- exit code: 9009

Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

Check C file whitespace..................................................Failed
- hook id: c-file-whitespace
- exit code: 9009

Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

Sphinx lint..............................................................Passed
Check hooks apply to the repository......................................Passed
Check for useless excludes...............................................Passed
```

**Comment by AA-Turner** [comment]
It isn't well documented, but I had hoped [``language: script``](https://pre-commit.com/#script) offered the salvation we desire. Sadly; no: 

```
Check Python file whitespace.............................................Failed
- hook id: python-file-whitespace
- exit code: 1

Executable `python3` not found

Check C file whitespace..................................................Failed
- hook id: c-file-whitespace
- exit code: 1

Executable `python3` not found
```

This is, I think, as the shebang line says ``#! /usr/bin/env python3`` -- if I change it to ``#! /usr/bin/env python``, it passes for me. I'm not sure how to write such a line to choose between `python3` or `python`, though.

A

**Comment by hugovk** [comment]
With:
```yaml
        name: "Check Python file whitespace"
        entry: 'Tools/patchcheck/reindent.py --nobackup --newline LF'
```
And:
```py
#! /usr/bin/env python
```
On macOS I get:
```
Executable `python` not found
```

**Comment by AlexWaygood** [comment]
@AA-Turner, in https://github.com/python/cpython/pull/109891#issuecomment-1755755952 you commented that using a Python-language hook (rather than a system-language hook, as we have currently) might slow us down unnecessarily, as pre-commit would have to create a venv before the hook would run. But in my experience, the Python-language hooks in the pre-commit-hooks package are plenty fast enough. I think pre-commit _might_ locally cache the venvs it creates, and reuse them in future invocations of the same hook?

**Comment by erlend-aasland** [comment]
@AlexWaygood, does 9501fab02da6da722b3cc4df3c8a55b89891b766 work on Windows?

**Comment by AlexWaygood** [comment]
> @AlexWaygood, does [9501fab](https://github.com/python/cpython/commit/9501fab02da6da722b3cc4df3c8a55b89891b766) work on Windows?

Argh, that's still a no :((

<details>

```
>pre-commit run --all-files
[INFO] Initializing environment for local.
[INFO] Installing environment for local.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Run Ruff on Lib/test/....................................................Passed
Run Ruff on Argument Clinic..............................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
Check Python file whitespace.............................................Failed
- hook id: python-file-whitespace
- exit code: 9009

Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

Check C file whitespace..................................................Failed
- hook id: c-file-whitespace
- exit code: 9009

Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

Sphinx lint..............................................................Passed
Check hooks apply to the repository......................................Passed
Check for useless excludes...............................................Passed
```

</details>

**Comment by erlend-aasland** [comment]
Sounds like it's a pre-commit bug; their docs says it should work on Windows: https://pre-commit.com/#python

**Comment by AlexWaygood** [comment]
> Sounds like it's a pre-commit bug; their docs says it should work on Windows: https://pre-commit.com/#python

The docs also say this, however, which isn't what we're doing:

> The hook repository must be installable via `pip install .` (usually by either setup.py or pyproject.toml). The installed package will provide an executable that will match the [entry](https://pre-commit.com/#hooks-entry) – usually through console_scripts or scripts in setup.py.

I'm guessing this is the issue here -- the `python`-language hooks in the `pre-commit-hooks` repo are all installed hooks that have their own entry points: https://github.com/pre-commit/pre-commit-hooks/blob/27dcd3fd1dc01d3fdbeb188edb54dddf3d964236/setup.cfg#L31

**Comment by erlend-aasland** [comment]
Then let's make those scripts pip installable.

**Comment by erlend-aasland** [comment]
Closing; see issue.

**Pull request closed without merging**